text/0000-feature-metadata.md

```diff
@@ -0,0 +1,275 @@
+-   `doc`: A markdown docstring describing the feature. Like with `#[doc(...)]`,
+    the first line will be treated as a summary.
+-   `deprecated`: This can be either a simple boolean, a string, or an object
+    with `since` and/or `note` keys. Cargo will warn downstream crates using this
+-   `public`: A boolean flag defaulting to `true` that indicates whether or not
+    downstream crates should be allowed to use this feature
+-   `unstable`: A boolean flag indicating that the feature is only usable with
+    nightly Rust, or is otherwise not API-stable
+
+If a downstream crate attempts to use the features `baz` and `qux`, they will
+see messages like the following:
+
+```
+warning: feature `quux` on crate `mycrate` is deprecated since version
+  1.2.3: "don't use this!"
+
+error: feature `baz` on crate `mycrate` is private and cannot be used by
+  downstream crates
+```
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+Validation and parsing of the new schema, described above, shoud be relatively
+straightforward. Each of the added keys is discussed individually in the
+following sections:
+
+## `requires`
+
+`requires` will take the place of the feature dependency array that currently
+exists. Semantics will remain unchanged.
+
+## `doc`
+
+`doc` is the most straightforward: it accepts markdown-flavored text, and should
+be thought of as the equivalent to a `#[doc(...)]` attribute. Like doc comments,
+the first line should be usable as a summary, though most use cases will only
+have one line.
+
+There is nothing in this RFC that cargo `must` do with this action, since it is
+mainly intended for the consumption of `rustdoc` or `docs.rs`. However, it can
+be used for general diagnostic information.
+
+## `deprecated`
+
+`deprecated` should be thought of as the equivalent of the [`deprecated`]
+attribute in Rust source. The value can be a boolean, string, or an object with
+`since` or `note` keys. Schema rules are as follows:
+
+-   If a boolean value, `false` indicates not deprecated and `true` indicates
+    deprecated
+-   If an object, the keys `since` and/or `note` can be specified
+-   An empty object is not allowed to avoid the ambiguous
+    `foo = { deprecated = {} }`
+-   A string `foo = { deprecated = "my msg" }` will be equivalent to if that string
+    was specified in the `note` field:
+    `foo = { deprecated = { note = "my msg" } }`
+-   If not specified, the default is `false`
+
+If a downstream crate attempts to use a feature marked `deprecated`, Cargo
+should produce a warning.
+
+## `public`
+
+`public` is a boolean value that defaults to `true`. It can be thought of as
+`pub` in Rust source files, with the exception of being true by default. If set
+to `false`, Cargo should forbid its use with an error message on any downstream
+crates.
+
+There needs to be an escape hatch for this for things like benchmarks - RFC TBD
+on how this works.
```

Comment:
How come?

Reply:
I mentioned the reasons in https://github.com/rust-lang/rfcs/pull/3416#discussion_r1172915429, but I'm happier removing the `unstable` key and allowing use of nonpublic features if explicitly specified https://github.com/rust-lang/rfcs/pull/3416#discussion_r1172889980

Reply:
Not seeing a use case described in that comment but a re-iteration of what I'm asking about.

What does this have to do with removing `unstable` key?

Reply:
My initial thought was that `unstable` would likely be used for internal features that you want to test directly with benchmarks, integration tests (same crate), or a separate testing crate. At least that has been my use case in the past: I want to benchmark some internal function or test with some internal type in my crate but don't want it to be public API, so I hide it behind a feature flag and use [`visibility`](https://crates.io/crates/visibility). 

After some clarification here though, it seems like `public` would be a better flag to use for use case than `unstable` - which is what I meant to describe via my first link. However, this only fully works as long as there is a way to only warn/allow feature visibility checks (assuming they are deny by default), which, to my understanding, the discussion in the second link would provide.

Reply:
I think this and other problems could be solved by changing the rule from

>  If set to false, Cargo should forbid its use with an error message on any downstream crates.

to something like:

>  If set to false, Cargo should forbid its use with an error message except when the crate is a member of a workspace (i.e. not downloaded from a registry or `git`-dependency); in that case, the feature may be mentioned:
>
> * Explicitly in the build command (`cargo build --features private_feature`).
> * In a dependency; the dependent must also be local and have `[package] publish = false` to ensure that it does not break when published.
> * In `required-features` or [`enable-features`](https://github.com/rust-lang/rfcs/pull/3374) of a `[[target]]` in the same package.

Thus, the private feature can be used for all purposes related to developing the package, even in a complex multi-package workspace/repo.

Reply:
Note that this conversation is split between here and the next conversation.

In the next one, I mentioned the idea of using `public = false` features for marking a crate that can't be depended on for production (in TOML) but can be used on the command-line, like `clap/debug`.  Not sure if we should support it in the initial release or put it down as a future possibility.

As for what should or shouldn't be included, I'm hesitant about allowing access to non-public features within a workspace as we don't allow similar access within other parts of the language, like `pub` and the orphan rules.  What if we punted on this to a future possibility instead and we start with a more strict subset for `public`?

Reply:
Since this is marked as outdated but the next conversation isn't, I will resolve this one and follow up there (https://github.com/rust-lang/rfcs/pull/3416#discussion_r1172657530)